acl/acl.go

```diff
@@ -0,0 +1,115 @@
+// Package acl provides access to Fastly ACLs.
+//
+// See the [Fastly ACL documentation] for details.
+//
+// [Fastly ACL documentation]: https://www.fastly.com/documentation/guides/concepts/edge-state/dynamic-config/#access-control-lists
+package acl
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/fastly/compute-sdk-go/internal/abi/fastly"
+)
+
+var (
+	// ErrNotFound indicates the requested ACL was not found.
+	ErrNotFound = errors.New("acl: not found")
+
+	// ErrInvalidHandle indicatest the ACL handle was invalid.
+	ErrInvalidHandle = errors.New("acl: invalid handle")
+
+	// ErrInvalidResponseBody indicates the looup response body was invalid.
+	ErrInvalidResponseBody = errors.New("acl: invalid response body")
+
+	// ErrInvalidArgument indicates the IP address was invalid.
+	ErrInvalidArgument = errors.New("acl: invalid argument")
+
+	// ErrNoContent indicates there was no entry for the provided IP address.
+	ErrNoContent = errors.New("acl: no content")
+
+	// ErrTooManyRequests indicates too many requests were made.
+	ErrTooManyRequests = errors.New("acl: too many requests")
+
+	// ErrUnexpected indicates an unexpected error occurred.
+	ErrUnexpected = errors.New("acl: unexepected error")
+)
+
+// Handle is a handle for an ACL
+type Handle struct {
+	h *fastly.ACLHandle
+}
+
+// Response is an ACL lookup response
+type Response struct {
+	Prefix string // Matching prefix in CIDR notation
+	Action string // Associated prefix's action
+}
+
+// Open returns a handle to the named ACL.
+func Open(name string) (*Handle, error) {
+	a, err := fastly.OpenACL(name)
+	if err != nil {
+		status, ok := fastly.IsFastlyError(err)
+		switch {
+		case ok && status == fastly.FastlyStatusNone:
+			return nil, ErrNotFound
+		case ok:
+			return nil, fmt.Errorf("%w (%s)", ErrUnexpected, status)
+		default:
+			return nil, err
+		}
+	}
+
+	return &Handle{h: a}, nil
+
+}
+
+// Lookup the given IP in the ACL and returns the response.  If no match was found, returns ErrNoContent.
+func (h *Handle) Lookup(ip net.IP) (Response, error) {
+	body, err := h.h.Lookup(ip)
+	if err != nil {
+		return Response{}, mapFastlyErr(err)
+	}
+
+	var r Response
+	dec := json.NewDecoder(body)
+	if err := dec.Decode(&r); err != nil {
+		return Response{}, err
+	}
+	return r, nil
+}
+
+func mapFastlyErr(err error) error {
+	// Is it a acl-specific error?
+	if aclErr, ok := err.(fastly.ACLError); ok {
+		switch aclErr {
+		case fastly.ACLErrorUninitialized: // we really shouldn't be returning this
+			return ErrUnexpected
+		case fastly.ACLErrorOK:
+			// Not an error; we shouldn't get here
+			return fmt.Errorf("%w (%s)", ErrUnexpected, err)
+		case fastly.ACLErrorNoContent:
+			return ErrNoContent
```

Comment:
In the context of `Handle.Lookup`, I wonder if something like `ErrNoMatch` would be more easily understood?

Reply:
So replace `ErrNoContent` with a new error `ErrNoMatch` which improves the semenatic match but pulls it farther away from the underlying response code (and Rust SDK I guess).

Reply:
Yes, agree that it pulls it further from the underlying response code. The Rust SDK uses `Option<Response>` with `None` when there's no match. 

I think `ErrNoContent` is fine, but wanted to ask in case we could think of anything better.

Reply:
The other way to represent "not found" would be to return a pointer to the entry, or a nil pointer (and no error) if it's not found.   That puts the onus on the user, but they're already checking for error (and in fact maybe this is "nicer" since it doesn't conflate an error and a success-not-found case.)

Reply:
Yes, I think using a pointer/nil could work and agree that it's not conflating errors with not-found cases. One final alternative would be to return something like:
`(r Result, ok bool, err error)`
 where ok is false in the not-found case.

Reply:
Rethinking this; I think I'm going to stick with returning an error.  For example, Go's `database/sql` package returns `ErrNoRows` for query returning no rows.  This is the same.   And I want to stick with the error name in the wit file for consistency (and also in case things change in the future we won't end up in a situation where our SDK code drifts *more* from the underlying host calls.)